Fix variables in __all__ being flagged by --no-implicit-reexport

bluetech opened the pull request:
The natural expectation is that when a variable is explicitly added to a
module's `__all__`, the module explicitly means to export it.

Previously, this was not the case, and a `from [..] import [..] as [..]`
reexport was required by `--no-implicit-export` even if the variable is
listed in `__all__`.

Fixes https://github.com/python/mypy/issues/7042

bluetech commented:
Following @JukkaL suggestion in https://github.com/python/mypy/issues/7042#issuecomment-506690942, I tried to look at the code for a couple of hours and this is what I came up with. I am not sure it is correct at all, or if it has any unexpected consequences, so please review carefully (once/if it passes CI) :)

msullivan commented on mypy/newsemanal/semanal.py:
```diff
@@ -445,10 +445,13 @@ def add_builtin_aliases(self, tree: MypyFile) -> None:
             for name, g in self.globals.items():
-                if name not in self.all_exports:
+                if name in self.all_exports:
+                    g.module_hidden = False
```
Looking at the docs for `module_public` and `module_hidden`:
 * public means that it will be picked up by an `import *`
 * hidden means that it can't be imported at all

So I think we need to set `public = True` when the name does appear to prevent it from being spurious excluded from `import *`.

bluetech commented:
@msullivan Yes, testing with `*`-imports I see some unexpected things.

If I understand `module_public` and `module_hidden` correctly, this is what they mean (with `--no-implicit-reexport` enabled):

`module_public` | `module_hidden` | meaning
-----|-----------|-----------
false | false | cannot be imported directly, not included in `import *`
false | true | can be imported directly, not included in `import *`
true | false | cannot be imported directly, included in `import *`
true | true | can be imported directly, included in `import *`

But the "cannot be imported directly, included in import *" case, I think makes no sense? That is, I think that if a symbol is imported with `import *`, it should always be possible to import it directly.

If true, this suggests that instead of two booleans, an enum should be used, maybe something like this:

```python
class SymbolModuleVisibility:
    """The visibility of the symbol in a module, i.e. whether and how another module can import the symbol from this module.

    Only relevant for module symbol tables.
    """
    # Another module cannot import this symbol.
    hidden = 1
    # Another module can import this symbol directly.
    exported = 2
    # Another module can import this symbol directly, and with `from <module> import *`.
    exported_public = 3
```

Things which affect visibility (AFAICS) are:

- Name starts with `_`.
- Name appears in `__all__`.
- Name imported from another module with `as` (`from another_module import x as y`).
- Name is a module import without `as` (`import another_module`).
- Is stub file (per rules in [PEP 484 Stub Files section](https://www.python.org/dev/peps/pep-0484/#stub-files)).
- Implicit reexports are enabled.

Do you think it makes sense to get into that, or just do the fix you suggested and be done with it?

msullivan commented:
Removing the potential invalid state by using defined constants seems reasonable. (Unfortunately we can't use an `Enum` right now because of limitations in `mypyc`, a compiler we use to compile `mypy`)

bluetech commented:
I applied @msullivan's suggestion and also adjusted the docs.

I also added a second commit which switches the code to use an enum as proposed in the comment above.

I want to add more test cases and probably adjust the terminology in the second commit, so marking as WIP.

msullivan commented on mypy/nodes.py:
```diff
@@ -2824,6 +2824,20 @@ def accept(self, visitor: NodeVisitor[T]) -> T:
+    Only relevant for module symbol tables.
+    """
+    # Another module cannot import this symbol.
+    hidden = 1  # type: Final
```
Make these names all caps

bluetech commented:
Left a comment in the issue https://github.com/python/mypy/issues/7042#issuecomment-509005322 on what's blocking this.

ilevkivskyi commented:
@bluetech 
What is the status of this PR? It fixes a high-priority issue and looks mostly ready apart from a merge conflict caused by moving semantic analyzer files.

bluetech commented:
@ilevkivskyi Sorry for not following up. This needs some work/research, in particular spec'ing out the rules for each of these:

- runtime imports
- type checking regular files with implicit exports
- type checking regular files without implicit exports
- type checking stub files

per each type of import and export (partial list?)

- `import x`
- `import x as z`
- `import x.y` (what happens to `x`)
- `import x.y as z`
- `from x import y`
- `from x import y as z`
- `from x import *`
- `__all__` defined?
- symbol starts with `_`?

and I haven't had enough patience to do it. For now I'd like to close it, but instead I'll just try to go back to the initial commit and rebase which at least makes the `__all__` exports explicit, and you can decide if you want to merge it or close it.

bluetech commented:
Updated as described above.

ilevkivskyi commented:
I think we can keep this open (after https://github.com/python/mypy/pull/7361 is merged) as a place to discuss possible follow-up refactoring/clean-up that will replace `module_public`/`module_hidden` with an enum and clarifies the corner cases.

bluetech commented:
This was squashed in to e9cf4f2e797d76d68e895d9ecc39b1baa56dfa81 (PR #7361).
